### ISSUE TITLE
[BUG] `satisfies` method doesn't differentiate between different prerelease identifiers

### ISSUE THREAD

**Issue by dayton-bobbitt** [npm/node-semver#331]
# What
My team maintains an npm package that is under active development. At any given time, we have multiple releases available for our users to consume (e.g. `2.x.x@latest`, `3.0.0-beta.x@beta`, and `3.0.0-canary.x@canary`).

Recently, we've noticed an issue where npm is unable to differentiate between releases with different prerelease identifiers, and npm will install a canary release when it should install a beta release. 

Essentially, we want to be able to publish multiple prerelease versions, some for public use and some for internal use, and easily document how teams can consume the latest version of a given release (e.g. always install the latest release of `beta`). I know that we could instruct teams to manually replace `^3.0.0-beta.0` with `beta` as the version in their package.json file but it'd be nice if there was a way to do this automatically with the `npm install` command when they run `npm install --save ourPackage@beta`. My guess is that this is a limitation / design decision of the `semver` package, and more specifically the `semver.satisfies` method.

*I fully understand that this may be a misunderstanding on my part of how npm and semver resolve versions and how prereleases are supposed to work. If that's the case, how should I be approaching this issue instead?*

## When
* Whenever the package.json references a prerelease version (e.g. `^3.0.0-beta.0`) and another prerelease with a different identifier has been published more recently (e.g. `3.0.0-canary.1`)
* Whenever evaluating the `semver.satisfies` method with versions using different prerelease identifiers (e.g. `semver.satisfies("3.0.0-canary.0", "^3.0.0-beta.0"); // returns true`)

## Where
* npm/cli
* npm/node-semver

### Current Behavior
1. Consuming team wants to install our beta release. They run `npm install --save ourPackage@beta` and their package.json file is updated to include `"ourPackage": "^3.0.0-beta.x"`
2. My team publishes a canary release for internal testing with the distribution tag `canary` (e.g. `npm publish --tag canary`)
3. In the consuming team's build pipeline, they run `npm install` and the version of `ourPackage` that gets installed is `3.0.0-canary.x` instead of our latest `beta` release.

### Steps to Reproduce
The easiest way to reproduce is with the `semver` package:
```javascript
const semver = require("semver");

semver.satisfies("3.0.0-canary.0", "^3.0.0-beta.0");  // returns true
```

### Expected Behavior
* I would expect `"ourPackage": "^3.0.0-beta.0"` to install the latest beta release of our package
* I would expect `semver.satisfies("3.0.0-canary.0", "^3.0.0-beta.0")` to return `false` because it's comparing two different prereleases published with two different distribution tags.

## Who
* n/a

## References
* n/a

**Comment by matheo** [comment]
and I would expect `semver.satisfies('16.0.0-beta-0', '*')` to return `true`
because the definition of the range is `"" := "*" := >=0.0.0` (Any version satisfies)

**Comment by dayton-bobbitt** [comment]
Bump

**Comment by dayton-bobbitt** [comment]
Bump

**Comment by ljharb** [comment]
What if you pass the `includePrelease: true` option?

**Comment by matheo** [comment]
@ljharb a lot of OpenSource projects out threre trusts `semver` to work out-of-the-box with the documented logic
`lerna` does that and is currently broken because `semver` is not comparing prereleases properly...

**Comment by ljharb** [comment]
@matheo sure, i'm not saying that's the solution, i'm asking to ensure that that's the reason it's not working.

**Comment by matheo** [comment]
@ljharb unfortunately I don't have the time right now to retry the tests.
we've downgraded the version to `semver@6` like the major projects out there, and works fine.

**Comment by dayton-bobbitt** [comment]
Hi @ljharb, thanks for the reply! Is this the correct syntax?

```javascript
const semver = require("semver");

semver.satisfies("3.0.0-canary.0", "^3.0.0-beta.0", {
    includePrerelease: true,
});
```

If so, this is unfortunately still returning `true` when I would've expected `false`.

**Comment by ljharb** [comment]
No, `-canary` and `-beta` have no relationship to each other. You wayt `^3.0.0-canary.0`.

**Comment by dayton-bobbitt** [comment]
I think `-canary` and `-beta` having no relationship to one another actually helps describe the bug I'm documenting. `semver.satisfies(...)` should not return `true` because there *is no relation* but it *is* returning `true`.

**Comment by dayton-bobbitt** [comment]
And just to clarify, I'm not personally using the `semver` package; I'm experiencing this issue with `npm` itself but I think it may be caused by this issue.

**Comment by Emptymu** [comment]
Any update on this?

**Comment by Emptymu** [comment]
> and I would expect `semver.satisfies('16.0.0-beta-0', '*')` to return `true` because the definition of the range is `"" := "*" := >=0.0.0` (Any version satisfies)

Any update on this?

**Comment by lukekarrys** [comment]
> I think -canary and -beta having no relationship to one another actually helps describe the bug I'm documenting. semver.satisfies(...) should not return true because there is no relation but it is returning true.

I thought this too (that's why I made #405) but I learned from the spec that pre-release versions are sorted lexically if the identifier is just letters. From https://semver.org/#spec-item-11:

```
Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:

  1. Identifiers consisting of only digits are compared numerically.

  2. Identifiers with letters or hyphens are compared lexically in ASCII sort order.

  3. Numeric identifiers always have lower precedence than non-numeric identifiers.

  4. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.

Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.
```

So it conforms to the spec that `3.0.0-canary.0 > 3.0.0-beta.0`.

If you want a semver range that only includes `beta` versions and not `canary`, I'd recommend something like `>=3.0.0-beta.0 <3.0.0-beta.z`.

```js
> s.satisfies('3.0.0-beta.0', ">=3.0.0-beta.0 <3.0.0-beta.z")
true
> s.satisfies('3.0.0-canary.0', ">=3.0.0-beta.0 <3.0.0-beta.z")
false
```
